### PR TITLE
IL Editing GenPasses

### PR DESCRIPTION
**Pull request by Destructor-Ben**
### What is the new feature?

Implementation of #3110 

Adds `WorldGen.ModifyPass` and `WorldGen.DetourPass` to allow easily IL hooking legacy world gen passes

### Why should this be part of tModLoader?

Obtaining the `MethodBase` for the world gen delegates previously required using a temporary read-only IL hook and scanning the method for the reference. Now it can be pulled directly from the delegate.

### Sample usage for the new feature

```cs
// All of the registration should take place in load
// Generation pass hooks are unloaded manually, so no Unload method is needed
public override void Load() {
	// IL editing the pyramids pass
	WorldGen.ModifyPass((PassLegacy)WorldGen.VanillaGenPasses["Pyramids"], Modify_Pyramids);

	// Detouring the shinies pass (generates ore)
	WorldGen.DetourPass((PassLegacy)WorldGen.VanillaGenPasses["Shinies"], Detour_Shinies);
}
```

### ExampleMod updates
See `ExampleWorldGenHookingSystem`

### CODE REVIEW

File: patches/tModLoader/Terraria/WorldGen.TML.cs [hunk 1]
```diff
@@ -7,4 +19,80 @@ public partial class WorldGen
...
+			HookEndpointManager.Unmodify(GetGenPassMethod(ilEdit.Item1), ilEdit.Item2);
+		}
+
+		foreach (var detour in genPassDetours) {
+			HookEndpointManager.Remove(GetGenPassMethod(detour.Item1), detour.Item2);
+		}
+	}
+
+	// TODO: tidy up
+	internal static MethodInfo GetGenPassMethod(GenPass task)
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/WorldGen.TML.cs]: Recommend this and `DetourTask` above take a `PassLegacy`. Let the compiler push the requirement onto the user so they're aware.

File: patches/tModLoader/Terraria/WorldGen.TML.cs [hunk 2]
```diff
@@ -7,4 +19,80 @@ public partial class WorldGen
...
+			}
+		}
+		else {
+			throw new Exception("Task is not a PassLegacy");
+			return null;
+		}
+	}
+
+	// TODO: Remove after testing and create DumpHooks for detours
+	internal static void DumpILHooks()
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/WorldGen.TML.cs]: This could be useful in a separate PR. A better place to access this info is in `MonoModHooks.cs`?

File: patches/tModLoader/Terraria/WorldGen.TML.cs [hunk 3]
```diff
@@ -7,4 +19,80 @@ public partial class WorldGen
...
+	/// <summary>
+	/// IL edits a PassLegacy during world generation. Use in ModifyWorldGenerationTasks or ModifyHardmodeTasks.
+	/// </summary>
+	/// <param name="task">The task to be IL edited</param>
+	/// <param name="callback">The method that IL edits the pass</param>
+	// TODO: Add support for hardmode tasks
+	public static void ModifyTask(GenPass task, ILContext.Manipulator callback)
+	{
+		HookEndpointManager.Modify(GetGenPassMethod(task), callback);
+		genPassILEdits.Add(Tuple.Create(task, callback));
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/WorldGen.TML.cs]: I don't like the idea of adding and removing IL hooks after load time. It 'should' be safe, but it's a fair bit of overhead, repeat logging, more risk of errors etc.

Fortunately, 1.4.4 made it so that none of the legacy gen pass delegates access any local variables anymore! That means you can do what we call a 'method split'

Break the `WorldGen.GenerateWorld` method into 3 methods, `GenerateWorld`, `AddGenPasses` and `GenerateWorld_RunTasksAndFinish`. 

```cs
		AddGenPasses();
		GenerateWorld_RunTasksAndFinish(seed, customProgressObject, generationStopwatch);
	}

	internal static void AddGenPasses() {
		AddGenerationPass("Reset", delegate (GenerationProgress progress, GameConfiguration passConfig) {
```

later
```cs
	private static void RunTasks(int seed, GenerationProgress customProgressObject, Stopwatch generationStopwatch) {
		SystemLoader.ModifyWorldGenTasks(_generator._passes, ref _generator._totalLoadWeight);
	
	...
```

Then you can call something like:
```cs
_generator = new();
AddGenPasses();
SystemLoader.HookGenPasses(_generator._passes)
```

Honestly, it doesn't look like any of the gen passes are conditionally added, so you could probably just store them all in a list at load, and expose it as `IReadOnlyList<GenPass> VanillaGenPasses` in `WorldGen.TML.cs`

Then your example could just use `WorldGen.EditGenPass("Living Trees", il=> ...`

**Review by Chicken-Bones — CHANGES_REQUESTED** [review]

File: patches/tModLoader/Terraria/WorldGen.TML.cs (hunk 2 above)
Comment (Destructor-Ben) [reply patches/tModLoader/Terraria/WorldGen.TML.cs]: This was mainly so I could tell that the hooks were being registered and unregistered properly, but I can move it to https://github.com/tModLoader/tModLoader/issues/3213

File: patches/tModLoader/Terraria/WorldGen.TML.cs (hunk 3 above)
Comment (Destructor-Ben) [reply patches/tModLoader/Terraria/WorldGen.TML.cs]: Would IL editing other mods generation passes be important? I'm guessing not, since it would probably be a smaller group of people wanting it, and I'm guessing most mods would use GenPasses instead of PassLegacys, but in case it is, should all world generation passes get moved to being added at load time (including modded ones)?

Comment (Chicken-Bones) [reply patches/tModLoader/Terraria/WorldGen.TML.cs]: IL editing requires intimate knowledge of the target code. Presumably, modded gen passes would have their code in one or more methods, not just a lambda.

Either way, I wouldn't bother considering mods IL editing mods when making tML decisions

File: patches/tModLoader/Terraria/WorldGen.cs.patch [hunk 4]
```diff
@@ -1,5 +1,13 @@
 --- src/TerrariaNetCore/Terraria/WorldGen.cs
 +++ src/tModLoader/Terraria/WorldGen.cs
+@@ -4,6 +_,7 @@
+ using System.Threading;
+ using System.Threading.Tasks;
+ using Microsoft.Xna.Framework;
++using rail;
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/WorldGen.cs.patch]: accidental import?

File: patches/tModLoader/Terraria/WorldGen.TML.cs [hunk 5]
```diff
@@ -7,4 +23,34 @@ public partial class WorldGen
 	private static double _timePass = 0.0; // Used to account for more precise time rates.
 
 	internal static void ClearGenerationPasses() => _generator?._passes.Clear();
+
+	internal static List<GenPass> vanillaGenPasses_internal = new();
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/WorldGen.TML.cs]: Preferred name is `_vanillaGenPasses`

File: patches/tModLoader/Terraria/WorldGen.TML.cs [hunk 6]
```diff
@@ -7,4 +23,34 @@ public partial class WorldGen
...
+	public static ReadOnlyCollection<GenPass> VanillaGenPasses = vanillaGenPasses_internal.AsReadOnly();
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/WorldGen.TML.cs]: prefer to return `IReadOnlyCollection`. Sure someone can cast it back to `List` but we're not here stop that.

File: ExampleMod/Common/Systems/ExampleWorldGenHookingSystem.cs [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+{
+	// This ModSystem will demonstrate how to IL edit and Detour world generation passes
+	// Since world generation passes are anonymous methods (they don't have a name), they can't be edited the standard way (using IL_xx or On_xx)
+	public class ExampleWorldGenHookingSystem : ModSystem
+	{
+		// All of the registration should take place in load
+		// Generation pass hooks are unloaded manually, so no Unload method is needed
+		public override void Load() {
+			// IL editing the pyramids pass
+			GenPass pyramidsPass = WorldGen.VanillaGenPasses.First(t => t.Name == "Pyramids");// Locate the generation pass
```
Comment (Chicken-Bones) [inline ExampleMod/Common/Systems/ExampleWorldGenHookingSystem.cs]: Need to call `FirstOrDefault` here if you want it to be error free

Comment (Chicken-Bones) [reply ExampleMod/Common/Systems/ExampleWorldGenHookingSystem.cs]: Honestly, I'd prefer using `Single` and a cast, with no warning. Since no-one should be modifying the vanilla pass list anyway.

**Review by Chicken-Bones — CHANGES_REQUESTED** [review]
Looking good

File: ExampleMod/Common/Systems/ExampleWorldGenHookingSystem.cs [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+{
+	// This ModSystem will demonstrate how to IL edit and Detour world generation passes
+	// Since world generation passes are anonymous methods (they don't have a name), they can't be edited the standard way (using IL_xx or On_xx)
+	public class ExampleWorldGenHookingSystem : ModSystem
+	{
+		// All of the registration should take place in load
+		// Generation pass hooks are unloaded manually, so no Unload method is needed
+		public override void Load() {
+			// IL editing the pyramids pass
+			GenPass pyramidsPass = WorldGen.VanillaGenPasses.FirstOrDefault(t => t.Name == "Pyramids");// Locate the generation pass
```
Comment (Chicken-Bones) [inline ExampleMod/Common/Systems/ExampleWorldGenHookingSystem.cs]: Final feedback, on thinking about this more. 
- `VanillaGenPasses` should be an `IDictionary<string, GenPass>`
- Examples should just access the value, expecting it to exist, since nothing can (or should) remove vanilla gen passes.

**Comment by Chicken-Bones** [comment]
I think the only minor change remaining on this is my `try catch` recommendation for the examples. I'm happy to contribute it and finish this if you'd like. 

Also, please send me a ping on the tModLoader Discord, I'd love to chat with you there!

**Comment by Destructor-Ben** [comment]
> I'm happy to contribute it and finish this if you'd like

That would be really nice, thanks

File: patches/tModLoader/Terraria/WorldGen.TML.cs [hunk 9]
```diff
@@ -1,4 +1,12 @@
+using MonoMod.Cil;
+using MonoMod.RuntimeDetour.HookGen;
+using System.Collections.Generic;
+using System.Reflection;
+using System;
+using Terraria.GameContent.Generation;
 using Terraria.WorldBuilding;
+using Terraria.IO;
+using System.Collections.ObjectModel;
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/WorldGen.TML.cs]: Imports cleanup?
